dmoj/cptbox/_cptbox.pyx

```diff
@@ -588,7 +587,25 @@ cdef class Process:
     def signal(self):
         if not self._exited:
             return None
-        return self._signal if self.process.was_initialized() else 0
+
+        if not self.process.was_initialized():
+            return None
+
+        # Some runtimes may use signals internally (e.g. Haskell's ticker
+        # thread sends `SIGVTALRM`), in which case a successful execution would
+        # have a nonzero `_signal`. This would lead the judge into displaying
+        # confusing output (e.g., "IR (virtual timer expired)").
+        #
+        # Avoid such cases. While we can't for sure distinguish signals that
+        # killed a process from signals that were part of normal execution, a
+        # zero or positive exit code is a good proxy.
+        #
+        # A more aggressive check here would be to ensure `self._exitcode ==
+        # -self.signal`.
```

Comment:
What do you mean you can't know for sure? Non-negative exit codes imply process didn't exit due to a signal, and this is guaranteed by `cptbox`: https://github.com/DMOJ/judge-server/blob/dbd148900ac7d052b5afd751ab0413cec72295a8/dmoj/cptbox/ptproc.cpp#L352